src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LiveShare/RemoteHierarchyService.cs

```diff
@@ -4,48 +4,48 @@
 using System;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor;
 using Microsoft.VisualStudio.LiveShare;
 using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Shell.Interop;
 using Microsoft.VisualStudio.Threading;
 
 namespace Microsoft.VisualStudio.Razor.LiveShare;
 
-internal class RemoteHierarchyService(CollaborationSession session, JoinableTaskFactory jtf) : IRemoteHierarchyService
+internal sealed class RemoteHierarchyService(
+    CollaborationSession session,
+    IVsService<SVsUIShellOpenDocument, IVsUIShellOpenDocument> vsUIShellOpenDocumentService,
+    JoinableTaskFactory jtf)
+    : IRemoteHierarchyService
 {
+    private readonly IVsService<SVsUIShellOpenDocument, IVsUIShellOpenDocument> _vsUIShellOpenDocumentService = vsUIShellOpenDocumentService;
     private readonly CollaborationSession _session = session;
     private readonly JoinableTaskFactory _jtf = jtf;
 
     public async Task<bool> HasCapabilityAsync(Uri pathOfFileInProject, string capability, CancellationToken cancellationToken)
     {
-        if (pathOfFileInProject is null)
-        {
-            throw new ArgumentNullException(nameof(pathOfFileInProject));
-        }
+        ArgHelper.ThrowIfNull(pathOfFileInProject);
+        ArgHelper.ThrowIfNull(capability);
 
-        if (capability is null)
+        var vsUIShellHostDocument = await _vsUIShellOpenDocumentService.GetValueOrNullAsync(cancellationToken);
```

Comment:
Just confirming, this used to switch to the UI thread before getting the service. It's intended that changed?

Reply:
Yeah, it's OK. Before, it was calling `IServiceProvider.GetService(...)` and casting the result to `IVsUIShellOpenDocument`. Because `IVsUIShellOpenDocument` is a COM STA-bound interface, that would cause an implicit `JTF.Run(...)` in order to perform the implicit `QueryInterface` call needed to cast on the UI thread. So, it was good practice to switch to the UI thread before requesting the service.

In this case, [`IVsService<,>.GetValueOrNullAsync(...)` calls into `IAsyncServiceProvider.GetServiceAsync<,>(...)`](https://devdiv.visualstudio.com/DevDiv/_git/VS?path=/src/env/shell/UIInternal/Packages/AsyncQueryService/VsServiceExporter.cs&version=GBmain&line=55&lineEnd=56&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) which will request the service and use a non-blocking switch to the UI thread to QI for the COM interface. This is a safer way to do this. See [the DevDiv Wiki](https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki/24027/Threading?anchor=%E2%9D%8C-**avoid**-blindly-casting-the-result-of-iasyncserviceprovider.getserviceasync) for more details.